backstop/src/emissions/claim.rs

```diff
@@ -13,26 +14,63 @@ pub fn execute_claim(e: &Env, from: &Address, pool_addresses: &Vec<Address>, to:
     }
 
     let mut claimed: i128 = 0;
+    let mut claims: Map<Address, i128> = Map::new(e);
     for pool_id in pool_addresses.iter() {
         let pool_balance = storage::get_pool_balance(e, &pool_id);
         let user_balance = storage::get_user_balance(e, &pool_id, from);
-        claimed += update_emissions(e, &pool_id, &pool_balance, from, &user_balance, true);
+        let claim_amt = update_emissions(e, &pool_id, &pool_balance, from, &user_balance, true);
+
+        claimed += claim_amt;
+        claims.set(pool_id, claim_amt);
     }
 
     if claimed > 0 {
-        let blnd_token = TokenClient::new(e, &storage::get_blnd_token(e));
-        blnd_token.transfer(&e.current_contract_address(), to, &claimed);
+        let blnd_id = storage::get_blnd_token(e);
+        let lp_id = storage::get_backstop_token(e);
+        TokenClient::new(e, &blnd_id).approve(
+            &e.current_contract_address(),
+            &lp_id,
+            &claimed,
+            &(e.ledger().sequence() + 10),
+        );
+        let lp_tokens_out = CometClient::new(e, &lp_id).dep_tokn_amt_in_get_lp_tokns_out(
+            &blnd_id,
+            &claimed,
+            &0,
+            &e.current_contract_address(),
+        );
+        for pool_id in pool_addresses.iter() {
+            let claim_amount = claims.get(pool_id.clone()).unwrap();
+            let deposit_amount = lp_tokens_out
+                .fixed_mul_floor(claim_amount, claimed)
+                .unwrap();
+            let mut pool_balance = storage::get_pool_balance(e, &pool_id);
+            let mut user_balance = storage::get_user_balance(e, &pool_id, to);
+
+            // Deposit LP tokens into pool backstop
+            let to_mint = pool_balance.convert_to_shares(deposit_amount);
+            pool_balance.deposit(deposit_amount, to_mint);
+            user_balance.add_shares(to_mint);
+
+            storage::set_pool_balance(e, &pool_id, &pool_balance);
+            storage::set_user_balance(e, &pool_id, to, &user_balance);
+            e.events().publish(
+                (Symbol::new(&e, "deposit"), pool_id, to),
+                (deposit_amount, to_mint),
+            );
```

Comment:
not a massive fan of this logic being duplicated, but there doesn't seem to be that easy of a way around it without a bunch of refactoring to make things consistent